nomad/core_sched.go

```diff
@@ -344,16 +343,20 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
 	return gcEval, gcAllocIDs, nil
 }
 
-// olderVersionTerminalAllocs returns terminal allocations whose job create index
-// is older than the job's create index
-func olderVersionTerminalAllocs(allocs []*structs.Allocation, job *structs.Job) []string {
+// olderVersionTerminalAllocs returns a tuplie ([]string, bool). The first element is the list of
+// terminal allocations which may be garbage collected for batch jobs. The second element indicates
+// whether or not the allocation itself may be garbage collected.
+func olderVersionTerminalAllocs(allocs []*structs.Allocation, job *structs.Job, thresholdIndex uint64) ([]string, bool) {
 	var ret []string
+	var mayGCEval = true
 	for _, alloc := range allocs {
-		if alloc.Job != nil && alloc.Job.CreateIndex < job.CreateIndex && alloc.TerminalStatus() {
+		if alloc.CreateIndex < job.JobModifyIndex && alloc.ModifyIndex < thresholdIndex && alloc.TerminalStatus() {
```

Comment:
Are these checks enough to ensure that the following scenario doesn't exhibit?

1. Batch job submitted
2. Eval created
3. Alloc created
4. Alloc finishes
5. Batch job updated
6. Is it possible for no new Eval/Alloc pair to correspond to the update?

Reply:
I think so. `(*Job).JobModifyIndex` is only updated when the user-facing version of the job is updated (as opposed to `(*Job).ModifyIndex` which is updated whenever the job is updated in raft for any reason, like being marked complete or whatever.